File: TankWorld.Code/Common/TankWorld.Common/Map/Percolation.vb
Imports TankWorld.UnionFind

<Obsolete("Please use the F# code", True)>
Public Class Percolation_VB
    Dim uf As WeightedQuickUnion
    Dim width As Integer
    Dim height As Integer
    Dim openSites() As Boolean

    ''' <summary>
    ''' Creates a rectangular with width * height, the first block is  (0,0)
    ''' </summary>
    ''' <param name="width">the width of the blocks, starting from 0</param>
    ''' <param name="height">the height of the blocks, starting from 0</param>
    Public Sub New(width As Integer, height As Integer)
        If (width <= 0 Or height <= 0) Then
            Throw New ArgumentOutOfRangeException()
        End If

        Me.width = width
        Me.height = height
        uf = New WeightedQuickUnion(width * height)

        openSites = New Boolean(width * height) {}
    End Sub

    ''' <summary>
    ''' true if the site is open
    ''' </summary>
    ''' <param name="x">the x value</param>
    ''' <param name="y">the y value</param>
    ''' <returns>true if the site is open</returns>
    Public Function isOpen(x As Integer, y As Integer) As Boolean
        Return openSites(x + y * Me.width)
    End Function

    Private Function GetOneDimensionIndex(x As Integer, y As Integer) As Integer
        Return x + y * Me.width
    End Function

    ''' <summary>
    ''' Open site x, y
    ''' </summary>
    ''' <param name="x">x</param>
    ''' <param name="y">y</param>
    Public Sub Open(x As Integer, y As Integer)
        ValidateHeight(y)
        ValidateWidth(x)

        If openSites(x + y * Me.width) Then
            Return
        End If
        openSites(x + y * Me.width) = True

        If x <> 0 Then
            UnionIfOpen(x + y * Me.width, x - 1 + y * Me.width)
        End If
        If x <> Me.width - 1 Then
            UnionIfOpen(x + y * Me.width, x + 1 + y * Me.width)
        End If

        If y <> 0 Then
            UnionIfOpen(x + y * Me.width, x + (y - 1) * Me.width)
        End If

        If y <> Me.height - 1 Then
            UnionIfOpen(x + y * Me.width, x + (y + 1) * Me.width)
        End If

    End Sub

    Private Sub UnionIfOpen(p As Integer, q As Integer)
        If Not openSites(q) Then
            Return
        End If
        uf.Union(p, q)
    End Sub

    Private Sub ValidateWidth(index As Integer)
        If (index < 0 Or index >= Me.width) Then
            Throw New ArgumentOutOfRangeException("index", "index " & index & " is not between 0 and " & (Me.width - 1))
        End If
    End Sub

    Private Sub ValidateHeight(index As Integer)
        If (index < 0 Or index >= Me.height) Then
            Throw New ArgumentOutOfRangeException("index", "index " & index & " is not between 0 and " & (Me.height - 1))
        End If
    End Sub

    ''' <summary>
    ''' true if point(x1, y1) and point(x2, y2) are connected
    ''' </summary>
    ''' <param name="x1">x value of the first point</param>
    ''' <param name="y1">y value of the first point</param>
    ''' <param name="x2">x value of the second point</param>
    ''' <param name="y2">y value of the second point</param>
    ''' <returns>true if point(x1, y1) and point(x2, y2) are connected</returns>
    Public Function Connected(x1 As Integer, y1 As Integer, x2 As Integer, y2 As Integer) As Boolean
        Return uf.Connected(x1 + y1 * width, x2 + y2 * width)
    End Function

End Class

File: TankWorld.Code/Common/TankWorld.UnionFind/Percolation.fs
(*
Created By Weiguang Zhou - Oct 25, 2015

This class creates a m*n rectangular and checks if two points are connected.
It uses WeightedQuickUnion class.

*)


namespace TankWorld.UnionFind

/// <summary>
/// Creates a rectangular with width * height, the first block is  (0,0)
/// </summary>
/// <param name="width">the width of the blocks, starting from 0</param>
/// <param name="height">the height of the blocks, starting from 0</param>
type Percolation(width, height) =

    let uf = new WeightedQuickUnion(width*height)
    let mutable openSites = [for i in 1..(width*height) -> false]
    
    //in order to use WeightedQuickUnion, we need to transform the two dimension coordinates to one dimension index.
    let oneDimensionIndex x  y  =
        x + y * width

             //validate that p Is a valid index
    let validateWidth x=
        if x>=width || x<0  then raise(System.IndexOutOfRangeException("index " + x.ToString() + " is not between 0 and " + (width - 1).ToString()))
        |> ignore

             //validate that p Is a valid index
    let validateHeight y=
        if y>=height || y<0  then raise(System.IndexOutOfRangeException("index " + y.ToString() + " is not between 0 and " + (height - 1).ToString()))
        |> ignore

    let unionIfOpen (p:int) (q:int) =
        if ( openSites.Item(q)) then uf.Union(p,q)
        |> ignore

    let rec changeItemInner headTemp list item index  =
        match list with
        |[]->[]
        |h::t->if index=0 then headTemp @ [item] @ t else changeItemInner (headTemp@[h]) t item (index-1) 

    let rec changeItem = changeItemInner []

    /// <summary>
    /// true if the site is open
    /// </summary>
    /// <param name="w">the x value</param>
    /// <param name="h">the y value</param>
    /// <returns>true if the site is open</returns>
    member t.isOpen( w, h) = openSites.Item(w + h * width)

    /// <summary>
    /// Open site x, y
    /// </summary>
    /// <param name="x">x</param>
    /// <param name="y">y</param>
    member t.Open( w, h) =
        validateWidth(w)
        validateHeight(h)
        let oneDim=oneDimensionIndex w h
        if not(openSites.Item(oneDim )) then
            //  [for i in 0 .. width*height-1 -> if i=oneDim then true else openSites.Item(i)]
            openSites <- changeItem openSites  true oneDim 
            if not (w = 0) then 
                unionIfOpen oneDim (oneDim - 1)
            if not (w = width-1) then 
                unionIfOpen oneDim (oneDim + 1)
            if not (h = 0) then 
                unionIfOpen oneDim (oneDim - width)
            if not (h = height - 1) then 
                unionIfOpen oneDim (oneDim + width)
        |> ignore

    /// <summary>
    /// true if point(x1, y1) and point(x2, y2) are connected
    /// </summary>
    /// <param name="x1">x value of the first point</param>
    /// <param name="y1">y value of the first point</param>
    /// <param name="x2">x value of the second point</param>
    /// <param name="y2">y value of the second point</param>
    /// <returns>true if point(x1, y1) and point(x2, y2) are connected</returns>
    member t.Connected x1 y1 x2 y2 = 
        uf.Connected(oneDimensionIndex x1 y1, oneDimensionIndex x2 y2)



